File: POH3/after-time-answer.js
// paiza POH! Lite
// result:
//
// author: Leonardone @ NEETSDKASU
process.stdin.resume();
process.stdin.setEncoding('utf8');
process.stdin.on('data', function (chunk) {
    var lines = chunk.toString().split('\n');
    var m = lines[0] ^ 0;
    var n = lines[1] ^ 0;
    var w = -m;
    var p = 0;
    var i,j,line,x,y,u,v;
    for(i=n+1; i>1; i--) {
        line = lines[i] = lines[i].split(" ");
        w += line[0] ^= 0;
        p += line[1] ^= 0;
    }
    var a = new Array(w+1);
    var b = new Array(w+1);
    var c = a;
    for(j=0; j<=w; j++) {
        a[j] = 0;
    }
    for(i=0; i<n; i++) {
        line=lines[i+2];
        u=line[0];
        v=line[1];
        if (u <= w) {
            for(j=0; j<u; j++) {
                b[j] = a[j];
            }
            for(; j<=w; j++) {
                x = a[j];
                y = a[j-u] + v;
                b[j] = (x > y) ? x : y;
            }
        } else {
            for(j=0; j<=w; j++) {
                b[j] = a[j];
            }
        }
        c = a;
        a = b;
        b = c;
    }
    console.log(p - a[w]);
});

File: POH3/after-time-answer.java
/* paiza POH! Lite
* result:
*
* author: Leonardone @ NEETSDKASU
*/
import java.io.*;
import java.lang.*;
import java.util.*;

class Main
{
    static byte[] buf = new byte[2000];
    static int bufindex = 0;
   
    static int nextInt() {
        int n = 0;
        int b = (int)buf[bufindex];
        while (b < '0' || b > '9') {
            b = (int)buf[++bufindex];
        }
        while (b >= '0' && b <= '9') {
            n = n * 10 + (int)(b) - '0';
            b = (int)buf[++bufindex];
        }
        return n;
    }
   
    static int[][] a = new int[51][500001];
    static int[] q = new int[51];
    static int[] r = new int[51];

    public static void main (String[] args) throws java.lang.Exception
    {
       
        int m;
        int n;
       
        int w = 0, p = 0;
       
        int i, j, x, y;
        int[] t0, t1;
       
        System.in.read(buf);
       
        m = nextInt();
        n = nextInt();

        for (i = 0; i < n; i++) {
            q[i] = nextInt();
            r[i] = nextInt();
            w += q[i];
            p += r[i];
        }
        w -= m;
       
        for (i = 0; i < n; i++) {
            t0 = a[i];
            t1 = a[i + 1];
            for (j = 0; j <= w; j++) {
                if (q[i] <= j) {
                    x = t0[j];
                    y = t0[j - q[i]] + r[i];
                    t1[j] = (x > y) ? x : y;
                } else {
                    t1[j] = t0[j];
                }
            }
        }
       
        System.out.println(p - a[n][w]);

    }
}
